source/common/http/filter/squash_filter.cc

```diff
@@ -0,0 +1,340 @@
+            json_config->getObject("metadata", true)->getString("name", EMPTY_STRING);
+      } catch (Json::Exception&) {
+        debugAttachmentId = EMPTY_STRING;
+      }
+
+      if (debugAttachmentId.empty()) {
+        ENVOY_LOG(info, "Squash: failed to parse debug attachment object - check server settings.");
+        doneSquashing();
+      } else {
+        debugAttachmentPath_ = POST_ATTACHMENT_PATH + debugAttachmentId;
+        pollForAttachment();
+      }
+    }
+
+    break;
+  }
+  case State::CheckAttachment: {
+
+    std::string attachmentstate;
+    try {
+      Json::ObjectSharedPtr json_config = getJsonBody(std::move(m));
+      attachmentstate = json_config->getObject("status", true)->getString("state", EMPTY_STRING);
+    } catch (Json::Exception&) {
+      // no state yet.. leave it empty for the retry logic.
+    }
+
+    bool attached = attachmentstate == ATTACHED_STATE;
+    bool error = attachmentstate == ERROR_STATE;
+    bool finalstate = attached || error;
+
+    if (finalstate) {
+      doneSquashing();
+    } else {
+      retry();
+    }
+    break;
+  }
+  }
+}
+
+void SquashFilter::onFailure(AsyncClient::FailureReason) {
+  // in_flight_request_ will be null if we are called inline of async client send()
+  bool request_created = in_flight_request_ != nullptr;
+  in_flight_request_ = nullptr;
+  switch (state_) {
+  case State::Initial: {
+    // Should never happen.
+    ENVOY_LOG(info, "Squash: received send failure callback when no request is in progress");
+    break;
+  }
+  case State::CreateConfig: {
+    // no retries here, as we couldnt create the attachment object.
+    if (request_created) {
+      // cleanup not needed if onFailure called inline in async client send.
+      // this means that decodeHeaders is down the stack and will return Continue.
+      doneSquashing();
+    }
+    break;
+  }
+  case State::CheckAttachment: {
+    retry();
+    break;
+  }
+  }
+}
+
+void SquashFilter::retry() {
+
+  if (delay_timer_.get() == nullptr) {
+    delay_timer_ =
+        decoder_callbacks_->dispatcher().createTimer([this]() -> void { pollForAttachment(); });
+  }
+  delay_timer_->enableTimer(config_->attachmentPollPeriod());
+}
+
+void SquashFilter::pollForAttachment() {
+  MessagePtr request(new RequestMessageImpl());
+  request->headers().insertMethod().value().setReference(Headers::get().MethodValues.Get);
+  request->headers().insertPath().value().setReference(debugAttachmentPath_);
+  request->headers().insertHost().value().setReference(SERVER_AUTHORITY);
+
+  in_flight_request_ = cm_.httpAsyncClientForCluster(config_->clusterName())
+                           .send(std::move(request), *this, config_->requestTimeout());
+  // no need to check if in_flight_request_ is null as onFailure will take care of
+  // cleanup.
+}
+
+void SquashFilter::doneSquashing() {
+  cleanup();
+  decoder_callbacks_->continueDecoding();
+}
+
+void SquashFilter::cleanup() {
+  state_ = State::Initial;
```

Comment:
have you considered having a State::Done?

Reply:
The main reason I use the Initial state is because I couldn't get a clear understanding on whether or not the Filter object can be 'recycled' and reused on a different request stream, or is it one per stream.

Reply:
That totally makes sense.  It should be one per stream.